termbox/src/lib.rs

```diff
@@ -405,6 +383,57 @@ impl Termbox {
             .unwrap();
         }
     }
+
+    fn send_char(&mut self, to_x: u16, to_y: u16, ch: char) {
+        // if the target cell is next to the last cell, then don't move the cursor
+        if (self.last_cursor.0 + 1) == to_x && self.last_cursor.1 == to_y {
+            write!(self.output_buffer, "{}", ch).unwrap();
+        } else {
+            goto(to_x, to_y, &mut self.output_buffer);
+            write!(self.output_buffer, "{}", ch).unwrap();
+        }
+        self.last_cursor = (to_x, to_y);
+    }
+}
+
+const LOOKUP: &[u8] = b"0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+fn num_to_buf(mut num: u16, buf: &mut Vec<u8>) {
+    let start_len = buf.len();
+    let mut chars_len = 0;
+    loop {
+        let rem = num % 10;
+        let ch = LOOKUP[rem as usize];
+        buf.push(ch);
+        num /= 10;
+        chars_len += 1;
+        if num == 0 {
+            break;
+        }
+    }
+
+    if chars_len == 1 {
+        return;
+    }
+
+    let swap_len = start_len + (chars_len / 2) as usize;
+    let mut cur_char;
+    let mut c = 0;
+    for i in start_len..swap_len {
+        let next_swap_idx = start_len + chars_len - c - 1;
+        cur_char = buf[next_swap_idx];
+        buf[next_swap_idx] = buf[i];
+        buf[i] = cur_char;
+        c += 1;
+    }
+}
+
+fn goto(x: u16, y: u16, buf: &mut Vec<u8>) {
+    write!(buf, "{}", "\x1B[").unwrap();
+    num_to_buf(y, buf);
+    write!(buf, "{}", ';').unwrap();
+    num_to_buf(x, buf);
+    write!(buf, "{}", 'H').unwrap();
```

Comment:
Could you replace these `write!`s with `buf.extend_from_slice()`s? That's simpler (no magic macros) and you won't need to `unwrap()`.

Reply:
Sure. I suppose the other write! macros, for things like colors, are another issue?